core/controllers/async_download.py

```diff
@@ -0,0 +1,147 @@
+
+        # email the URL to the user
+        send_email(
+            template_id=email_template_id,
+            email_address=email_address,
+            download_url=url,
+            find_service_url=find_service_base_url + "/download",
+            file_format=file_format_str,
+            file_size_str=file_size_str,
+        )
+
+
+def upload_file(file: IO | FileStorage, bucket: str, object_name: str, expiry: int) -> str:
+    """Upload a file to an S3 bucket, generate presigned url and return it
+    :param file: File to upload,
+    :param bucket: Bucket to upload to,
+    :param object_name: S3 object name (filename),
+    :param expiry: URL expiry time in seconds,
+
+    :return: Presigned URL
+    """
+
+    _S3_CLIENT.put_object(Bucket=bucket, Key=object_name, Body=file)
+
+    url = _S3_CLIENT.generate_presigned_url(
+        "get_object",
+        Params={"Bucket": bucket, "Key": object_name},
+        ExpiresIn=expiry,
+    )
+
+    return url
+
+
+def send_email(
+    template_id: str, email_address: str, download_url: str, find_service_url: str, file_format: str, file_size_str: str
+):
+    """Send an email to the user with the download link.
+    :param template_id: NOTIFY email template ID,
+    :param email_address: user's email address,
+    :param download_url: download URL,
+    :param find_service_url: URL of the FIND service,
+    :param file_format: human-readable file format string,
+    :param file_size_str: human-readable file size,
+    """
+
+    try:
+        notifications_client = NotificationsAPIClient(current_app.config["NOTIFY_FIND_API_KEY"])
+    except KeyError:
+        current_app.logger.error("NOTIFY_FIND_API_KEY is not set.")
+    else:
+        notifications_client.send_email_notification(
+            email_address=email_address,
+            template_id=template_id,
+            personalisation={
+                "download_url": download_url,
+                "find_service_url": find_service_url,
+                "file_format": file_format,
+                "file_size": file_size_str,
+            },
+        )
+
+
+def get_file_format_from_extension(file_extension: str) -> str:
+    """Return nice file format name based on the file extension.
+    :param file_extension: file extension,
+    :return: nice file format name,
+    """
+
+    if file_extension == "xlsx":
+        file_format = "Microsoft Excel spreadsheet"
+    elif file_extension == "json":
+        file_format = "JSON file"
+    return file_format
+
+
+def get_human_readable_file_size(file_size_bytes: int) -> str:
+    """Return a human-readable file size string.
+    :param file_size_bytes: file size in bytes,
+    :return: human-readable file size,
+    """
+
+    file_size_kb = round(file_size_bytes / 1024, 1)
+    if file_size_kb < 1024:
+        return f"{round(file_size_kb, 1)} KB"
+    else:
+        return f"{round(file_size_kb / 1024, 1)} MB"
```

Comment:
Thanks for splitting these out into their own functions - nice 👍  Can we add a few simple tests for them? And for the `get_human_readable_file_size`. It's probably fine that this only does KB and MB but wouldn't be super against adding support for bytes and GB too :) Not compulsory though.

Reply:
Return option added for 'GB', unit tests have been added for both function.